File: web/src/utils/timer.ts
import type { Segment } from '../types/segment';
import type { Timer } from '../types/timer';

const zeroPad = (num: number, places: number) => String(num).padStart(places, '0');
const getTimerText = (millis: number) => {
	let seconds = Math.floor(millis / 1000);
	if (seconds < 0) {
		seconds = 0;
	}

	let remaningHours = zeroPad(Math.floor(seconds / 60 / 60) % (60 * 60), 2);
	let remaningMinutes = zeroPad(Math.floor(seconds / 60) % 60, 2);
	let remaningSeconds = zeroPad(seconds % 60, 2);

	if (parseInt(remaningHours) === 0) {
		return remaningMinutes + ':' + remaningSeconds;
	}

	return remaningHours + ':' + remaningMinutes + ':' + remaningSeconds;
};

function calculateTimeInCurrentRound(
	timerData: Timer,
	currentTime: number
): { timeInCurrentRound: number; state: 'running' | 'waiting' | 'finished' | 'stopped' } {
	let stopped = false;

	if (timerData.stop_at && timerData.stop_at < currentTime) {
		currentTime = timerData.stop_at;
		stopped = true;
	}

	const elapsedTime = currentTime - timerData.start_at;

	if (elapsedTime < 0) {
		return {
			timeInCurrentRound: 0,
			state: 'waiting'
		};
	}

	const totalTimePerRound = timerData.segments.reduce((acc, curr) => acc + curr.time, 0);

	if (!timerData.repeat && !stopped && elapsedTime > totalTimePerRound) {
		return {
			timeInCurrentRound: 0,
			state: 'finished'
		};
	}

	let timeInCurrentRound = elapsedTime % totalTimePerRound;
	return {
		timeInCurrentRound,
		state: stopped ? 'stopped' : 'running'
	};
}

function calculateTimeInCurrentSegment(
	timeInCurrentRound: number,
	segments: Segment[]
): {
	currentSegment: Segment;
	timeInCurrentSegment: number;
} {
	let currentSegmentIndex = 0;
	let timeInCurrentSegment = 0;
	while (timeInCurrentRound > 0) {
		timeInCurrentSegment = Math.floor(segments[currentSegmentIndex].time - timeInCurrentRound);
		timeInCurrentRound -= segments[currentSegmentIndex].time;
		currentSegmentIndex++;
	}

	const currentSegment = segments[currentSegmentIndex - 1];

	return {
		currentSegment,
		timeInCurrentSegment
	};
}

export { calculateTimeInCurrentRound, calculateTimeInCurrentSegment, getTimerText };

File: display/src/timer.cpp
#include <time.h>

#include "timer.h"

namespace timer {

TimerData _timerData;

TimerData *timerData() { return &_timerData; }

TIME calculateTimeInCurrentRound(TIME currentTime) {

  if (currentTime < _timerData.start_at &&
      _timerData.display_options.pre_start_behaviour ==
          PreStartBehaviour::SHOW_FIRST_SEGMENT) {
    return 1;
  }

  long totalTimePerRound = 0;
  for (int i = 0; i < 10 && _timerData.segments[i].valid; i++) {
    totalTimePerRound += _timerData.segments[i].time;
  }

  if (currentTime < _timerData.start_at &&
      _timerData.display_options.pre_start_behaviour ==
          PreStartBehaviour::SHOW_LAST_SEGMENT) {
    return totalTimePerRound;
  }

  long elapsedTime = currentTime - _timerData.start_at;

  if (_timerData.stop_at != 0 && currentTime > _timerData.stop_at) {
    elapsedTime = _timerData.stop_at - _timerData.start_at;
  }

  if (!_timerData.repeat && elapsedTime > totalTimePerRound) {
    return totalTimePerRound;
  }

  return elapsedTime % totalTimePerRound;
}

ActiveSegment calculateCurrentSegment(TIME timeOffset) {
  if (!_timerData.valid || timeOffset == 0) {
    return {0, 0xfff, "", 0};
  }

  TIME currentTime = (TIME)millis() + timeOffset;
  long timeInCurrentRound = calculateTimeInCurrentRound(currentTime);

  int currentSegmentIndex = 0;
  long timeInCurrentSegment = 0;
  while (timeInCurrentRound > 0 &&
         _timerData.segments[currentSegmentIndex].valid) {
    timeInCurrentSegment =
        _timerData.segments[currentSegmentIndex].time - timeInCurrentRound;
    timeInCurrentRound -= _timerData.segments[currentSegmentIndex].time;
    currentSegmentIndex++;
  }

  timeInCurrentSegment += _timerData.segments[currentSegmentIndex - 1].count_to;

  return {timeInCurrentSegment / 1000,
          _timerData.segments[currentSegmentIndex - 1].color,
          _timerData.segments[currentSegmentIndex - 1].label, currentTime};
}

}; // namespace timer